src/librustc_typeck/check/closure.rs

```diff
@@ -620,6 +637,117 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         result
     }
 
+    /// Invoked when we are translating the generator that results
+    /// from desugaring an `async fn`. Returns the "sugared" return
+    /// type of the `async fn` -- that is, the return type that the
+    /// user specified. The "desugared" return type is a `impl
+    /// Future<Output = T>`, so we do this by searching through the
+    /// obligations to extract the `T`.
```

Comment:
Maybe leave a FIXME/open an issue somewhere about generalizing this so it can rely entirely on the `impl`s involved and not have to know about the `Future` trait at all?